### PR TITLE
Introduce global disabling_analyzer() hook to veto disable_analyzer()

### PR DESCRIPTION
**Pull request by awelzel**
This hook can be used to coordinate disabling an analyzer for a given
connection. The contract is simple: Any script can veto a disable_analyzer()
call by breaking from this hook. The decision is local to the script taking
into account any state attached to the connection object or script specific
state stored elsewhere.
A script breaking from the hook takes over the responsibility to call
disable_analyzer() at a later point when it finds the condition due to which
it vetoed fulfilled (which may be never).

Signature:

    disabling_analyzer: hook(c: connection, atype: AllAnalyzers::Tag, aid: count);

Example use-cases are keeping the SSL analyzer enabled for finger-printing
until a certain amount of bytes or packets have been transferred or
similarly the connection duration exceed a certain threshold.

Other example use-cases might be keeping analyzers for SSH, RDP or SSL
enabled for connections from specific subnets.

It's a bit quirky as it makes disable_analyzer() a maybe operation. While log
policy hooks and/or the notice hook have similar semantics, they are not as
stateful. It still seems like a quite powerful primitive.

The disable_analyzer() call in dpd/main.zeek may motivate the addition of a
force flag as a follow-up for situations where the caller "knows better" or
absolutely wants to override.

Closes https://github.com/zeek/zeek/issues/1678 https://github.com/zeek/zeek/issues/1593.

---
Looking for comments on the approach, naming and semantics.

### CODE REVIEW

**Comment by awelzel** [comment]
@vpax - you had opened #1593 originally and @anthonykasza #1678. I believe they are very related. Would you (or someone working on the involved scripts) mind looking through the proposal and see if that could work out?

**Comment by awelzel** [comment]
> @awelzel Yes, this looks handy for the use cases we have. One thing a bit peculiar will be that if the analyzer is responsible for later calling `disable_analyzer()`, its hook will presumably _again_ be called, which means it may need to squirrel away some state just so it knows to decline that second call. I'm not sure there's a clean fix for that, though.

Thanks!

Yes, a script calling disable_analyzer() itself will have it's own hook invoked again. It then should allow (not break) rather than decline the second call though. But either way, think the pattern would be to use state on `connection` or in a table for declining and if there's none or state indicates it's okay to disable, don't break.

File: scripts/base/init-bare.zeek [hunk 1]
```diff
@@ -584,6 +584,22 @@ type fa_metadata: record {
 	inferred: bool &default=T;
 };
 
+## A hook taking a connection, analyzer tag and analyzer id that can be
+## used to veto disabling analyzers. This hook is invoked synchronously
+## during a :zeek:see:`disable_analyzer` call.
+##
```
Comment (rsmmr) [inline scripts/base/init-bare.zeek:590]: Worth adding an explicit note here on *how* to veto: by breaking.

**Review by rsmmr — APPROVED** [review]
Looks great. Just one comment nit, and please add a `NEWS` entry.

**Review by rsmmr — APPROVED** [review]

**Pull request merged**